src/python/pants/backend/python/goals/lockfile.py

```diff
@@ -232,7 +234,11 @@ async def generate_lockfile(
     initial_lockfile_digest_contents = await Get(
         DigestContents, Digest, poetry_export_result.output_digest
     )
-    metadata = LockfileMetadata(req.requirements_hex_digest, req.interpreter_constraints)
+    metadata = LockfileMetadata.new(
+        req.requirements_hex_digest,
+        req.interpreter_constraints,
+        {Requirement.parse(i) for i in req.requirements},
```

Comment:
I'm not absolutely convinced that this is the best place to put this parsing logic. It may be worth moving it over to `new` at some point, but for now, this works.

Reply:
Adding an error in here. As mentioned on the other location where I'm using `.parse`, the exception that gets raised doesn't play nice with mypy, and that's going to get in the way of improving this. I'm going to return to this in the future.